src/sentry/relay/config/metric_extraction.py

```diff
@@ -1,26 +1,122 @@
 from dataclasses import dataclass
-from typing import Any, List, Literal, Sequence, Tuple, TypedDict, Union, cast
+from typing import Any, List, Literal, Optional, Sequence, Tuple, TypedDict, Union, cast
 
+from typing_extensions import NotRequired
+
+from sentry import features
 from sentry.api.endpoints.project_transaction_threshold import DEFAULT_THRESHOLD
+from sentry.constants import DataCategory
+from sentry.incidents.models import AlertRule, AlertRuleStatus
 from sentry.models import (
     Project,
     ProjectTransactionThreshold,
     ProjectTransactionThresholdOverride,
     TransactionMetric,
 )
 
+# GENERIC METRIC EXTRACTION
+
+
+# Name component of MRIs used for custom alert metrics.
+# TODO: Move to a common module.
+CUSTOM_ALERT_METRIC_NAME = "transactions/alert"
+
+# Version of the metric extraction config.
+_METRIC_EXTRACTION_VERSION = 1
+
 
-class RuleConditionInner(TypedDict):
-    op: Literal["eq", "gt", "gte"]
+RuleCondition = Union["LogicalRuleCondition", "ComparingRuleCondition", "NotRuleCondition"]
```

Comment:
Is this the most readable way to make a union of types? Does rule inheritance (more types) or [`Protocol`](https://docs.python.org/3/library/typing.html#typing.Protocol) ([PEP 544](https://peps.python.org/pep-0544/)) (fewer types) make things easier to read?

Reply:
Since most of these types are extremely similar and separate types do not give us a benefit on typed dicts, I'd opt for readability with fewer types. 

How would you propose to write it?

Reply:
Maybe with `Protocol` but not strongly opinionated. My suggestion comes from me not liking types coming from strings this way.